content/automate/ManageIQ/Cloud/Orchestration/Retirement/StateMachines/Methods.class/__methods__/common.rb

```diff
@@ -0,0 +1,46 @@
+#
+# Description: common methods needed whilst retirement.
+#
+# /ManageIQ/Cloud/Orchestration/Retirement/StateMachines/Methods.class/__methods__/start_retirement.rb
+module ManageIQ
+  module Automate
+    module Cloud
+      module Orchestration
+        module Retirement
+          module StateMachines
+            module Methods
+              def self.get_stack(handle)
+                # From $evm.object.
+                if (stack = handle.object['orchestration_stack'])
+                  handle.log(:info, 'Fetched stack from $evm.object["orchestration_stack"]')
+                  stack
+                elsif (stack_id = handle.object['orchestration_stack_id'].to_i) && stack_id != 0 &&
+                      (stack = handle.vmdb('orchestration_stack', stack_id))
+                  handle.log(:info, 'Fetched stack from $evm.object["orchestration_stack_id"]')
+                  stack
+
+                # From $evm.root.
```

Comment:
Looks like there is a nice opportunity to refactor this `object` and `root` lookup code.  I would create another method and pass the object reference (`handle.root` or `handle.object`) and maybe a string to help with the logging.

Reply:
@gmcculloug yeah was a bit more involved, but I think it's worth it. Only the diff is ugly now, sorry about that. So what I did was:

- move the mixin into `/ManageIQ/Cloud/Orchestration/Lifecycle` class because methods like `get_stack` and `get_service` should probably be consumed by other instances as well, not just Retirement
- opt-in for `include` syntax to get rid of the gigantic namespace in front of `get_stack` & `get_service`
- convert direct methods into class methods. Or else rubocop has hard time with include in global namespace
- fix the Travis: was failing because the mixin wasn't loaded

Looking forward to your opinion.